### PR TITLE
`evaluate` method for `GroupedReductionOp`

### PR DESCRIPTION
**Pull request by Priya2698**
Issue #670

### CODE REVIEW

File: csrc/ir/nodes.cpp [hunk 1]
```diff
@@ -1534,6 +1534,43 @@ int GroupedReductionOp::getExprIndexOfOutput(Val* output_val) const {
       false, "Not an output, ", output_val->toString(), ", of ", toString());
 }
 
+std::vector<PolymorphicValue> GroupedReductionOp::evaluate(
+    const ExpressionEvaluator& ee,
+    const std::vector<PolymorphicValue>& inputs) const {
+  std::vector<PolymorphicValue> grouped_reduction_out;
```
Comment (wujingyue) [inline csrc/ir/nodes.cpp:1540]:
```suggestion
  std::vector<PolymorphicValue> grouped_reduction_out;
  grouped_reduction_out.reserve(numHorizontallyGroupedExprs());
```

**Review by wujingyue — APPROVED** [review]

File: csrc/ir/nodes.cpp [hunk 2]
```diff
@@ -1534,6 +1534,44 @@ int GroupedReductionOp::getExprIndexOfOutput(Val* output_val) const {
...
+    std::vector<int64_t> reduction_axes;
+    for (const auto id : c10::irange(int64_t(out_i->getRootDomain().size()))) {
+      auto ax = out_i->getRootDomain().at(id);
+      if (ax->isReduction()) {
+        reduction_axes.push_back(id);
+      }
+    }
+    switch (getReductionOpType(i)) {
+      case BinaryOpType::Add:
+        grouped_reduction_out.emplace_back(at::sum(input_i, reduction_axes));
```
[the remark this answers is not in the document]
Comment (Priya2698) [reply csrc/ir/nodes.cpp:1558]: I considered that when implementing the evaluate method for `ReductionOp`. However, it seemed that these are not used from the ops at `arith.cpp`https://github.com/NVIDIA/Fuser/blob/f3ff377f83e4900e61fc62c9c4bc1a2b8053aa56/csrc/ops/arith.cpp#L1324. So we can keep it simple for now.

File: csrc/ir/nodes.cpp [hunk 3]
```diff
@@ -1534,6 +1534,44 @@ int GroupedReductionOp::getExprIndexOfOutput(Val* output_val) const {
       false, "Not an output, ", output_val->toString(), ", of ", toString());
 }
 
+std::vector<PolymorphicValue> GroupedReductionOp::evaluate(
+    const ExpressionEvaluator& ee,
+    const std::vector<PolymorphicValue>& inputs) const {
+  std::vector<PolymorphicValue> grouped_reduction_out;
+  grouped_reduction_out.reserve(numHorizontallyGroupedExprs());
+  for (const auto i : c10::irange(numHorizontallyGroupedExprs())) {
+    const auto& input_i = inputs.at(i).as<at::Tensor>();
```
[the remark this answers is not in the document]
Comment (Priya2698) [reply csrc/ir/nodes.cpp:1543]: That's a good convention to follow! Switched to _tensor and _tv suffix.

File: csrc/ir/nodes.cpp [hunk 4]
```diff
@@ -1534,6 +1534,44 @@ int GroupedReductionOp::getExprIndexOfOutput(Val* output_val) const {
...
+    const auto out_i = output(i)->as<TensorView>();
+    NVF_ERROR(
+        !out_i->hasRFactor(),
+        "Evaluation for rFactored reductions is not supported.");
+
+    std::vector<int64_t> reduction_axes;
+    for (const auto id : c10::irange(int64_t(out_i->getRootDomain().size()))) {
```
[the remark this answers is not in the document]
Comment (Priya2698) [reply csrc/ir/nodes.cpp]: Clangtidy complains about the type.

**Pull request merged**